packages/artifact/src/upload-specification.ts

```diff
@@ -0,0 +1,87 @@
+import * as fs from 'fs'
+import {debug} from '@actions/core'
+import {join, normalize} from 'path'
+import {checkArtifactName} from './utils'
+
+export interface UploadSpecification {
+  absoluteFilePath: string
+  uploadFilePath: string
+}
+
+/**
+ * Creates a specification that describes how each file that is part of the artifact will be uploaded
+ * @param artifactName the name of the artifact being uploaded. Used during upload to denote where the artifact is stored on the server
+ * @param rootDirectory an absolute file path that denotes the path that should be removed from the beginning of each artifact file
+ * @param artifactFiles a list of absolute file paths that denote what should be uploaded as part of the artifact
+ */
+export function getUploadSpecification(
+  artifactName: string,
+  rootDirectory: string,
+  artifactFiles: string[]
+): UploadSpecification[] {
+  checkArtifactName(artifactName)
+
+  const specifications: UploadSpecification[] = []
+
+  if (!fs.existsSync(rootDirectory)) {
+    throw new Error(`Provided rootDirectory ${rootDirectory} does not exist`)
+  }
+  if (!fs.lstatSync(rootDirectory).isDirectory()) {
+    throw new Error(
+      `Provided rootDirectory ${rootDirectory} is not a valid directory`
+    )
+  }
+  rootDirectory = normalize(rootDirectory)
+  /*
+     Example to demonstrate behavior
+     
+     Input:
+       artifactName: my-artifact
+       rootDirectory: '/home/user/files/plz-upload'
+       artifactFiles: [
+         '/home/user/files/plz-upload/file1.txt',
+         '/home/user/files/plz-upload/file2.txt',
+         '/home/user/files/plz-upload/dir/file3.txt'
+       ]
+     
+     Output:
+       specifications: [
+         ['/home/user/files/plz-upload/file1.txt', 'my-artifact/file1.txt'],
+         ['/home/user/files/plz-upload/file1.txt', 'my-artifact/file2.txt'],
+         ['/home/user/files/plz-upload/file1.txt', 'my-artifact/dir/file3.txt']
+       ]
+  */
+  for (let file of artifactFiles) {
+    if (!fs.existsSync(file)) {
+      throw new Error(`File ${file} does not exist`)
+    }
+
+    if (!fs.lstatSync(file).isDirectory()) {
+      file = normalize(file)
```

Comment:
Without `resolve` we can't guarantee this is absolute path, can we?

Reply:
There is a method that checks if a path is absolute, I don't use it anywhere though: https://nodejs.org/api/path.html#path_path_isabsolute_path

We specify that both `files` and `rootDirectory` are supposed to be absolute, but I think think my code will work if everything is relative as well. The only time there could be an issue is if one is absolute and the other is relative so `resolve` should be used to make sure the correct paths are being compared

Reply:
I've updated it so that `normalize` and `resolve` are used for all paths. FYI @joshmgross this strays a little from the ADR where we were requiring absolute file paths, we can use relative file paths now as well

Reply:
If you normalize `file`, you should normalize `rootDirectory` as well.  We need to be consistent, otherwise you will run into issues if you try to upload files outside of your `cwd`.

Reply:
Sorry I missed the line above.  This looks good.

Reply:
Yup, both `file` and `rootDirectory` get normalized